lib/src/source_visitor.dart

```diff
@@ -3539,33 +3563,57 @@ class SourceVisitor extends ThrowingAstVisitor {
 
   /// Emit a single mandatory newline.
   void newline() {
-    builder.writeWhitespace(Whitespace.newline);
+    builder.writeNewline();
   }
 
   /// Emit a two mandatory newlines.
   void twoNewlines() {
-    builder.writeWhitespace(Whitespace.twoNewlines);
+    builder.writeNewline(isDouble: true);
   }
 
   /// Allow either a single split or newline to be emitted before the next
   /// non-whitespace token based on whether a newline exists in the source
   /// between the last token and the next one.
   void splitOrNewline() {
-    builder.writeWhitespace(Whitespace.splitOrNewline);
+    if (_linesBeforeNextToken > 0) {
+      builder.writeNewline(nest: true);
+    } else {
+      split();
+    }
   }
 
   /// Allow either a single split or newline to be emitted before the next
-  /// non-whitespace token based on whether a newline exists in the source
+  /// non-whitespace token based on whether any blank lines exist in the source
   /// between the last token and the next one.
   void splitOrTwoNewlines() {
-    builder.writeWhitespace(Whitespace.splitOrTwoNewlines);
+    if (_linesBeforeNextToken > 1) {
+      twoNewlines();
+    } else {
+      split();
+    }
   }
 
   /// Allow either one or two newlines to be emitted before the next
-  /// non-whitespace token based on whether more than one newline exists in the
-  /// source between the last token and the next one.
+  /// non-whitespace token based on whether any blank lines exist in the source
+  /// between the last token and the next one.
   void oneOrTwoNewlines() {
-    builder.writeWhitespace(Whitespace.oneOrTwoNewlines);
+    if (_linesBeforeNextToken > 1) {
+      twoNewlines();
+    } else {
+      newline();
+    }
+  }
+
+  /// Whether there is a blank line between the last written token and the next
```

Comment:
[nit] Comment reads as if it were a bool. Maybe "The number of blank lines"

Reply:
Oops, good catch. Done.